File: examples/haskell/07-fp-features/main.hs
-- FP Features in Haskell
-- ========================
-- Pure functions, type classes, lazy evaluation, monads

import Data.List (nub, sort)
import Data.Char (toUpper)


-- ── Basic pure functions ──────────────────────────

double :: Int -> Int
double x = x * 2

square :: Int -> Int
square x = x * x


-- ── Higher-order functions ─────────────────────────

applyTwice :: (Int -> Int) -> Int -> Int
applyTwice f x = f (f x)

result1 :: Int
result1 = applyTwice (+ 1) 5  -- (5+1)+1 = 7

result2 :: Int
result2 = applyTwice (* 2) 3  -- (3*2)*2 = 12


-- ── Composition (.) ─────────────────────────────

shout :: String -> String
shout = map toUpper

greet :: String -> String
greet name = "Hello, " ++ name

greetLoud :: String -> String
greetLoud = shout . greet  -- (.) composes right-to-left: shout(greet(name))


-- ── Currying (partial application) ─────────────

add :: Int -> Int -> Int
add x y = x + y

addFive :: Int -> Int
addFive = add 5  -- Curried: waits for one argument


-- ── Map/Filter/Reduce ─────────────────────────

squaresOf :: [Int] -> [Int]
squaresOf xs = map square xs

evens :: [Int] -> [Int]
evens = filter even

sum :: [Int] -> Int
sum xs = foldl (+) 0 xs

product :: [Int] -> Int
product xs = foldl (*) 1 xs


-- ── List comprehension ───────────────────────────

squaresGT25 :: [Int] -> [Int]
squaresGT25 xs = [x*x | x <- xs, x*x > 25]


-- ── Pattern matching ─────────────────────────────

describe :: Int -> String
describe n
    | n < 0     = "negative"
    | n == 0    = "zero"
    | n == 1    = "one"
    | otherwise  = "many"


factorial :: Int -> Int
factorial 0 = 1
factorial n = n * factorial (n - 1)


fibonacci :: Int -> Int
fibonacci 0 = 0
fibonacci 1 = 1
fibonacci n = fibonacci (n - 1) + fibonacci (n - 2)


-- ── Lazy evaluation ─────────────────────────────

naturals :: [Int]
naturals = [1..]  -- Infinite list!

take10Naturals :: [Int]
take10Naturals = take 10 naturals

evensLazy :: [Int]
evensLazy = filter even naturals


-- ── Custom higher-order functions ─────────────────

allPredicate :: (a -> Bool) -> [a] -> Bool
allPredicate p xs = foldl (\acc x -> acc && p x) True xs

anyPredicate :: (a -> Bool) -> [a] -> Bool
anyPredicate p xs = foldl (\acc x -> acc || p x) False xs


-- ── Summary ─────────────────────────────────────

main :: IO ()
main = do

    putStrLn "--- Pure functions ---"
    putStrLn $ "double(5) = " ++ show (double 5)
    putStrLn $ "square(4) = " ++ show (square 4)

    putStrLn "\n--- Higher-order functions ---"
    putStrLn $ "applyTwice(+1, 5) = " ++ show result1
    putStrLn $ "applyTwice(*2, 3) = " ++ show result2

    putStrLn "\n--- Composition ---"
    putStrLn $ "greetLoud \"alice\" = " ++ show (greetLoud "alice")

    putStrLn "\n--- Currying ---"
    putStrLn $ "addFive 10 = " ++ show (addFive 10)

    putStrLn "\n--- Map/Filter/Reduce ---"
    let nums = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
    putStrLn $ "Squares: " ++ show (squaresOf nums)
    putStrLn $ "Evens: " ++ show (evens nums)
    putStrLn $ "Sum: " ++ show (sum nums)
    putStrLn $ "Product: " ++ show (product nums)

    putStrLn "\n--- List comprehension ---"
    putStrLn $ "Squares > 25: " ++ show (squaresGT25 nums)

    putStrLn "\n--- Pattern matching ---"
    putStrLn $ "describe(-5): " ++ describe (-5)
    putStrLn $ "describe(0): " ++ describe 0
    putStrLn $ "factorial(5): " ++ show (factorial 5)
    putStrLn $ "fibonacci(10): " ++ show (fibonacci 10)

    putStrLn "\n--- Lazy evaluation ---"
    putStrLn $ "First 10 naturals: " ++ show (take10Naturals)

    putStrLn "\n--- All/Any predicates ---"
    putStrLn $ "All evens in [2,4,6]: " ++ show (allPredicate even [2, 4, 6])
    putStrLn $ "Any even in [1,3,5]: " ++ show (anyPredicate even [1, 3, 5])

    putStrLn "\n--- Summary ---"
    putStrLn "Haskell FP features:"
    putStrLn "  - Pure functions: no side effects"
    putStrLn "  - Type inference: Hindley-Milner algorithm"
    putStrLn "  - Type classes: ad hoc polymorphism (Eq, Show, Num)"
    putStrLn "  - Higher-order: map, filter, fold, take, drop"
    putStrLn "  - Composition: (.) operator"
    putStrLn "  - Currying: partial application"
    putStrLn "  - Lazy evaluation: infinite lists, lazy streams"
    putStrLn "  - Pattern matching: case expressions"
    putStrLn "  - Recursion: primary control flow (no loops)"
    putStrLn "  - Monads: IO, Maybe, Either type classes"
    putStrLn "  - List comprehensions: [x*2 | x <- xs]"
    putStrLn "  - do-notation: monadic sequencing"

File: examples/erlang/07-fp-features/main.erl
%% FP Features in Erlang
%% =====================
%% Pure functions, immutable data, recursion, pattern matching

-module(main).
-export([main/0]).

main() ->
    %% ── Pure functions ─────────────────────────────

    io:format("--- Pure functions ---~n"),

    %% Function as variable
    Add = fun(A, B) -> A + B,
    io:format("Add(5, 3) = ~p~n", [Add(5, 3)]),

    %% Higher-order function
    ApplyTwice = fun(F, X) -> F(F(X)),
    io:format("ApplyTwice(fun(X)->X+1, 5) = ~p~n", [ApplyTwice(fun(X)->X+1, 5)]),

    %% Named function
    Square = fun(X) -> X * X,
    io:format("Square(4) = ~p~n", [Square(4)]).


    %% ── Higher-order list functions ─────────────────────

    io:format("~n--- Higher-order list functions ---~n"),

    Numbers = lists:seq(1, 10),

    %% map (lists:map/2)
    Squares = lists:map(fun(X) -> X * X, Numbers),
    io:format("Squares: ~p~n", [Squares]),

    %% filter (lists:filter/2)
    Evens = lists:filter(fun(X) -> X rem 2 =:= 0, Numbers),
    io:format("Evens: ~p~n", [Evens]),

    %% foldl (lists:foldl/3)
    Sum = lists:foldl(fun(Acc, X) -> Acc + X, 0, Numbers),
    io:format("Sum: ~p~n", [Sum]),

    %% find (lists:find/2)
    FirstEven = lists:find(fun(X) -> X rem 2 =:= 0, Numbers),
    io:format("First even: ~p~n", [FirstEven]),

    %% any (lists:any/2)
    HasOdd = lists:any(fun(X) -> X rem 2 =/= 0, Numbers),
    io:format("Has odd number? ~p~n", [HasOdd]),

    %% all (lists:all/2)
    AllLessThan20 = lists:all(fun(X) -> X < 20, Numbers),
    io:format("All < 20? ~p~n", [AllLessThan20]).


    %% ── List comprehension ─────────────────────────────

    io:format("~n--- List comprehension ---~n"),

    %% [X*X || X <- Lists, X > 5]
    SquaresGT25 = [X*X || X <- Numbers, X*X > 25],
    io:format("Squares > 25: ~p~n", [SquaresGT25]).

    %% [X || X <- Lists, X rem 2 =:= 0, X < 10]
    EvensLT10 = [X || X <- Numbers, X rem 2 =:= 0, X < 10],
    io:format("Evens < 10: ~p~n", [EvensLT10]).

    %% List comprehension with multiple generators
    Pairs = [{X, Y} || X <- [1, 2, 3], Y <- [4, 5]],
    io:format("Pairs: ~p~n", [Pairs]).


    %% ── Closures (anonymous functions) ─────────────────

    io:format("~n--- Closures ---~n"),

    MakeMultiplier = fun(Factor) ->
        fun(X) -> X * Factor
    end,

    TimesThree = MakeMultiplier(3),
    TimesFive = MakeMultiplier(5),

    io:format("TimesThree(7) = ~p~n", [TimesThree(7)]),
    io:format("TimesFive(8) = ~p~n", [TimesFive(8)]).


    %% ── Pattern matching in functions ─────────────────────

    io:format("~n--- Pattern matching ---~n"),

    %% Match in function heads
    DescribeNumber = fun
        (1) -> "one";
        (2) -> "two";
        (3) -> "three";
        (_) -> "many"
    end,

    io:format("Number 2: ~s~n", [DescribeNumber(2)]),
    io:format("Number 5: ~s~n", [DescribeNumber(5)]),
    io:format("Number 99: ~s~n", [DescribeNumber(99)]).


    %% ── Recursion ─────────────────────────────────────

    io:format("~n--- Recursion ---~n"),

    %% Factorial (tail-recursive)
    Factorial = fun
        (0) -> 1;
        (N) when N > 0 -> N * Factorial(N - 1)
    end,

    io:format("Factorial(5) = ~p~n", [Factorial(5)]),
    io:format("Factorial(10) = ~p~n", [Factorial(10)]).


    %% ── Tail-recursive sum (accumulator) ─────────────

    SumList = fun
        SumList([], Acc) -> Acc;
        SumList([H|T], Acc) -> SumList(T, Acc + H)
    end,

    io:format("SumList([1,2,3]) = ~p~n", [SumList([1, 2, 3])]).


    %% ── Lazy evaluation (only when requested) ────────────

    io:format("~n--- Lazy evaluation ---~n"),

    %% Infinite list (not evaluated until needed)
    Naturals = fun() -> naturals() end,
    %% naturals/0 returns infinite list: [0,1,2,...]

    TakeTen = naturals(),
    TenItems = lists:sublist(TakeTen, 0, 10),
    io:format("First 10 natural numbers: ~p~n", [TenItems]).


    %% ── Summary ─────────────────────────────────────

    io:format("~n--- Summary ---~n"),
    io:format("Erlang FP features:~n"),
    io:format("  - Pure functions: no side effects~n"),
    io:format("  - Immutable data: cannot be mutated~n"),
    io:format("  - Pattern matching: function clauses by pattern~n"),
    io:format("  - Recursion: only iteration mechanism~n"),
    io:format("  - Tail recursion: optimized by BEAM~n"),
    io:format("  - Higher-order: map, filter, foldl~n"),
    io:format("  - List comprehensions: [X || X <- Lists]~n"),
    io:format("  - Anonymous functions (funs): closures~n"),
    io:format("  - Lazy evaluation: only when requested~n"),
    io:format("  - No loops: use recursion instead~n").
